### PR TITLE
Return Failure early in `<*` and `*>` too

### PR DESCRIPTION
**Pull request by josephcsible**
### HLint

- [ ] I've changed the exposed interface (add new reexports, remove reexports, rename reexported things, etc.).
  - [ ] I've updated [`hlint.dhall`](https://github.com/kowainik/relude/blob/master/hlint/hlint.dhall) accordingly to my changes (add new rules for the new imports, remove old ones, when they are outdated, etc.).
  - [ ] I've generated the new `.hlint.yaml` file (see [this instructions](https://github.com/kowainik/relude#generating-hlintyaml)).

### General

- [x] I've updated the [CHANGELOG](https://github.com/kowainik/relude/blob/master/CHANGELOG.md) with the short description of my latest changes.
- [x] All new and existing tests pass.
- [x] I keep the code style used in the files I've changed (see [style-guide](https://github.com/kowainik/org/blob/master/style-guide.md#haskell-style-guide) for more details).
- [x] I've used the [`stylish-haskell` file](https://github.com/kowainik/relude/blob/master/.stylish-haskell.yaml).
- [ ] My change requires the documentation updates.
  - [ ] I've updated the documentation accordingly.
- [ ] I've added the `[ci skip]` text to the docs-only related commit's name.

### CODE REVIEW

File: src/Relude/Extra/Validation.hs [hunk 1]
```diff
@@ -155,17 +155,19 @@ instance Semigroup e => Applicative (Validation e) where
     {-# INLINE (<*>) #-}
 
     (*>) :: Validation e a -> Validation e b -> Validation e b
-    Failure el *> Failure er = Failure (el <> er)
-    Failure e  *> Success _  = Failure e
-    Success _  *> Failure e  = Failure e
-    Success _  *> Success b  = Success b
+    Failure e *> b = Failure $ case b of
+                                 Failure e' -> e <> e'
```
Comment (chshersh) [inline src/Relude/Extra/Validation.hs]: My understanding would be that implementing operators `*>` and `<*` manually is unnecessary. They both implemented in terms of `<*>` which already shortcuts early. Implementing them manually could be error-prone in the case of this instance and performance benefits are not obvious. So I think it's better to just remove the implementations for `*>` and `<*`

Comment (josephcsible) [reply src/Relude/Extra/Validation.hs]: Isn't it slightly faster to have all of these operators explicitly written, at least in the case where both sides are `Success`?

Comment (chshersh) [reply src/Relude/Extra/Validation.hs]: It _may be_ slightly faster. But GHC is very good at optimizing. And it's hard to predict how faster without benchmarking or looking at the core. On the other hand, the cost of the possible error and maintaining more code is higher. So, if we going to change code for `Validation`, it's better to wait for issue #176 to be implemented first to make sure that the refactoring is indeed valid.

**Comment by josephcsible** [comment]
I also added a unit test to ensure the short-circuiting works correctly, for all of the `Applicative` methods.

**Comment by josephcsible** [comment]
Consider this code:
```haskell
case thisWillFail *> thisWillTakeALongTime of
  Failure _ -> print "It didn't work."
  Success _ -> print "It worked!"
```

Since we know that the whole expression will be a `Failure` after the LHS fails, we now emit `Failure` itself sooner, so that `thisWillTakeALongTime` doesn't need to be evaluated unless/until the argument to `Failure` is evaluated.

**Comment by vrom911** [comment]
Sure, I do understand that, but I'm talking exactly about this PR. This feels like unchecked breaking change. @chshersh are you sure you would like to take such a risk? I'm personally using `Validation` in production, so I wouldn't hurry to get to the newer version of `relude` then.

**Comment by josephcsible** [comment]
Am I missing something? How is this a breaking change?

**Comment by chshersh** [comment]
@vrom911 Agree with you here. I'm also using `Validation` in production and it works. We have a separate issue to implement tests for `Validation`. So let's not change the code yet if it's untested.

**Comment by josephcsible** [comment]
Actually, this is worse than just a performance problem; we're not a valid [Applicative](https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Applicative.html#t:Applicative):

> The other methods have the following default definitions, which may be overridden with **equivalent** specialized implementations:
> 
> * `u *> v = (id <$ u) <*> v`
> * `u <* v = liftA2 const u v`

(emphasis mine) Ours aren't equivalent. If `u = Failure ["foo"]` and `v = undefined`, then the LHS of both of those evaluate to `undefined`, but the RHS of both evaluate to `Failure undefined`.

**Comment by josephcsible** [comment]
Now that #176 is fixed, is this ready to merge?

**Comment by chshersh** [comment]
@josephcsible Well, since we have tests now, we actually can implement our own more optimised versions of these operators (like it how was done at first). Because now we have more guarantees that they will work. And we won't rely on the inliner too much as a pleasant consequence.

**Comment by josephcsible** [comment]
@chshersh Done.

File: src/Relude/Extra/Validation.hs [hunk 2]
```diff
@@ -143,6 +145,9 @@ Success 8
 
 >>> liftA2 (+) a c
 Failure ["Not correct"]
+
+>>> [x | Success x <- [ha <*> e, c <* e, d *> e, liftA2 (+) d e]]
```
Comment (vrom911) [inline src/Relude/Extra/Validation.hs]: Heh, I wish these variable names could make a little sense, so the examples could become useful :cry: 
@chshersh should I create an issue for refactoring this?

Comment (chshersh) [reply src/Relude/Extra/Validation.hs]: @vrom911 Yes, please!

Comment (vrom911) [reply src/Relude/Extra/Validation.hs]: https://github.com/kowainik/relude/issues/221

**Pull request merged**